File: MPI/sequentialSortMPI.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <mpi.h>
#include <limits.h>

#define MASTER 0 // Master process
//#define END_OF_PROCESSING -2 // End of processing
#define INF INT_MAX

// COMPILAR
// mpicc sequentialSortMPI.c -o sequentialSortMPI -lm
// mpiexec -n 5 ./sequentialSortMPI

/*
IDEA: EL PROCESO master, GESTIONA EL PROCESO, CADA worker RECORRE EL ARRAY ENTERO CON SU ELEMENTO, 
Y COMPARA LOS DEMAS ELEMENTOS, SUMANDO A UN CONTADOR PARA QUE SE LO DEVUELVA AL master Y ESTE LO COLOQUE 
EN SU POSICION CORRECTA EN OTRO ARRAY.

*/

int* leeArchivo(int* n);
void printArray(int* a, int n);
int arrayOrdenado(int* a, int n); 

int main(int argc, char *argv[]){
    // Variables MPI
  	int myrank, tag, numProc,numWorkers;	// rank y tag de MPI y el numero de procesos creados (el primero es el master)
    MPI_Status status;			            // status para mas info 
                                                // (entre esta info esta el proceso que recibe al usar anysource)
    int END_OF_PROCESSING=-2;

    // Variable para calculo de tiempo de ejecucion
    double timeStart, timeEnd;	

    // Arrays dinamicos
    int* a;                         // Entrada
    int* b;                         // Salida
    // tamaño de los arrays 
    int n;                          
    
    // Variables del proceso MPI
    int arrayProc;                  // Puntero de la parte procesada del array a
    int pos=-1, val;                // Variables que se envian y reciben entre los procesos
    int cont;                       // Variable auxiliar para los workers, que cuentan los elementos mayores al recibido

	int i;      				    // Variable para los bucles
		 
    // Init
    tag = 1;
    srand(time(NULL));

    // Init MPI
    MPI_Init (&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myrank);	
    MPI_Comm_size(MPI_COMM_WORLD, &numProc); // Numero de procesos al ejecutar el programa
    numWorkers=numProc-1;

    // CONTROL DE ERRORES
    /*if (<Error>){
        if (myrank == MASTER) {
            printf("Wrong configuration for NROWS (%d). SIZE (%d) < numWorkers(%d)*NROWS(%d)\n", NROWS, SIZE, numProc-1, NROWS);
            printf ("At least, each worker must receive %d rows to be processed\n", NROWS);
        }
        MPI_Finalize();
        exit(-1);			
    }*/
		
    // master lee el archivo .txt con el array de entrada
    // cuenta el tamaño del array de entrada e inicializa el array de salida
    if (myrank == MASTER) {
        a=leeArchivo(&n); 

        // Init b
        b=(int *)malloc(n*sizeof(int));
        for(i=0;i<n;i++) b[i]=INF;

        // Envia el tamaño de los arrays a los workers
        MPI_Bcast (&n, 1, MPI_INT, MASTER, MPI_COMM_WORLD);  
        // Envia el array entero a los workers
        MPI_Bcast (a, n, MPI_INT, MASTER, MPI_COMM_WORLD);  
    }
    else {
        // Recibe el tamaño de los arrays
        MPI_Bcast (&n, 1, MPI_INT, MASTER, MPI_COMM_WORLD);         
        
        a = (int*)malloc (n*sizeof(int)); // Reserva memoria              
        // Recibe el array entero
        MPI_Bcast (a, n, MPI_INT, MASTER, MPI_COMM_WORLD); 
    }   
            
    // Comienza el timer una vez inicializado todo
    timeStart = MPI_Wtime();

    if (myrank == MASTER){
        // Init
        arrayProc=0;        

        // Distribucion inicial
        if(numWorkers<=n){        
            for (i=1;i<=numWorkers;i++) {            
                // Envia val
                MPI_Send(&a[arrayProc], 1, MPI_INT, i, tag, MPI_COMM_WORLD);
                // update
                arrayProc++;							
            }
        }
        else {
            int aux=numWorkers%n;
            for (i=1;i<=numWorkers-aux;i++) {            
                // Envia val
                MPI_Send(&a[arrayProc], 1, MPI_INT, i, tag, MPI_COMM_WORLD);
                // update
                arrayProc++;							
            }
            for(i=numWorkers-aux+1;i<=numWorkers;i++){
                MPI_Send(&END_OF_PROCESSING, 1, MPI_INT, i, tag, MPI_COMM_WORLD);
            }
            numWorkers-=aux;
        }

        // Hay mas elementos a procesar
        while (arrayProc < n) {		
            pos=-1;	
            // Recibe los resultados de algun worker
            MPI_Recv (&pos, 1, MPI_INT, MPI_ANY_SOURCE, tag, MPI_COMM_WORLD, &status);	            
            MPI_Recv (&val, 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD, &status);
            
            // Fin
            if(arrayProc==n) break;		

            // Procesa los datos recibidos
            // Si la posicion en el array b esta ocupado, es porque es el mismo valor
            // y busca el siguiente espacio disponible
            while(b[pos]!=INF) pos++;
            b[pos]=val;


            // Envia el siguiente valor a procesar       
            MPI_Send (&a[arrayProc], 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD);
            // Update
            arrayProc++;							
        }

        // Ultimos valores por procesar
        while(numWorkers--){
            MPI_Recv (&pos, 1, MPI_INT, MPI_ANY_SOURCE, tag, MPI_COMM_WORLD, &status);	            
            MPI_Recv (&val, 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD, &status);
            while(b[pos]!=INF) pos++;
            b[pos]=val;
            MPI_Send(&END_OF_PROCESSING, 1, MPI_INT, status.MPI_SOURCE, tag, MPI_COMM_WORLD); 
        }
        timeEnd = MPI_Wtime();
        printf("Tiempo de ejecucion: %f\n", timeEnd-timeStart);	
        if(arrayOrdenado(b,n)) printf("Array ordenado\n");
        else printf("Array no ordenado\n");
    }		
    else { // workers
        while(1) {
            // Recibe el valor del array a procesar
            MPI_Recv (&val, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD, &status);
            if(val==-2) break;

            cont=0;
            for(i=0;i<n;i++){
                if(a[i]<val) cont++;
            }
            
            MPI_Send(&cont, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD);
            MPI_Send(&val, 1, MPI_INT, MASTER, tag, MPI_COMM_WORLD);               
        }    
    }
		// End MPI environment
  		MPI_Finalize();

    return 0;
}



// cambiar el metodo de seleccionar .txt, usando argv[]
int* leeArchivo(int* n){
    int arrayTam=0;
    int capacidad=10;
    int* a=(int*)malloc(capacidad*sizeof(int));
    if (a == NULL) {
        perror("Error al asignar memoria. 1er");
        exit(EXIT_FAILURE);
    }

    FILE *archivo;
    char nombre_archivo[100];
    int archivoTam;      
    printf("Introduce el nombre del archivo: ");
    scanf("%s", nombre_archivo);
    archivoTam = strlen(nombre_archivo);
    strcat(nombre_archivo, ".txt");

    archivo = fopen(nombre_archivo, "r"); // Modo lectura
    if (archivo == NULL) {
        perror("Error al abrir el archivo");
        exit(EXIT_FAILURE);
    }

    int tmp;
    while (fscanf(archivo, "%d", &tmp) == 1) {        
        if (arrayTam == capacidad) {
            capacidad *= 2;
            a = (int*)realloc(a, capacidad * sizeof(int));
            if (a == NULL) {
                perror("Error al reasignar memoria");
                exit(EXIT_FAILURE);
            }
        }

        a[arrayTam++] = tmp;        
    }

    /*for (int i = 0; i < n; i++) {
        fscanf(archivo, "%d", &a[i]);
    }*/  
    fclose(archivo); 

    *n=arrayTam;
    return a;
}

void printArray(int* a, int n){
    printf("Array:\n");
    for (int i = 0; i < n; i++) {
        printf("%d ", a[i]);
    }
    printf("\n");
}

int arrayOrdenado(int* a, int n){
    for(int i=1;i<n;i++){
        if(a[i]!=a[i-1]+1)return 0;
    }
    return 1;
}

File: 1. MPI/PYTHON/Ordenaciones/Codigo_Explicado/sequentialSortMPI.py
from mpi4py import MPI
import sys
import os

# COMPILAR
# mpiexec -np 5 python prueba1.py

# Este programa tiene (np-1) workers, cada worker recibe un elemento
#   que envia el master para realizar el recorrido en el array

"""
COSTE O(N^2)

100000 10;70.1 - 30;56.0 - 15;56.3 - 25;
            MPI (20)    MPI(50)     NORMAL
10000       0.51                    4.5
100000      55          55.9        440.4
1000000    
"""






# PROBAR, AL ENVIAR DATOS, 
# - ES MAS EFICIENTE ENVIAR UNA PARTE DEL ARRAY
# - PONERLO COMO VARIABLE GLOBAL
# - ENVIAR TODO EL ARRAY CON MPI_Bcast()


def main():  	
    MASTER = 0              # int.
    END_OF_PROCESSING=-2
    INF=sys.maxsize

    timeStart=0.0           # double. Para medir el tiempo de ejecucion
    timeEnd=0.0
    
    
    # Arrays 
    a=[]                    # int[]. Array Entrada
    b=[]                    #        Salida Salida          
    n=0                     # int.   Tamaño de los arrays                           



    # Variables usadas en el proceso MPI    
    arrayProc=0             # int. Puntero de la parte procesada del array a
    pos=-1                  # int. Variables que se envian y reciben entre los procesos
    val=0                   
    cont=0                  # int. Variable auxiliar para los workers, que cuentan los elementos mayores al recibido

		    

    # Init MPI.  rank y tag de MPI y el numero de procesos creados (el primero es el master)
    tag=0
    comm=MPI.COMM_WORLD    
    status = MPI.Status()
    myrank=comm.Get_rank()
    numProc=comm.Get_size()
    numWorkers=numProc-1
   
    # CONTROL DE ERRORES
    """if (<Error>){
        if (myrank == MASTER) {
            printf("Wrong configuration for NROWS (%d). SIZE (%d) < numWorkers(%d)*NROWS(%d)\n", NROWS, SIZE, numProc-1, NROWS);
            printf ("At least, each worker must receive %d rows to be processed\n", NROWS);
        }
        MPI_Finalize();
        exit(-1);			
    }"""
		
    # master lee el archivo .txt con el array de entrada
    # cuenta el tamaño del array de entrada e inicializa el array de salida
    if myrank==MASTER: 
        a,n=leeArchivo() 
        b=[(INF) for i in range(n)]
    

    # Envia el tamaño de los arrays a los workers
    n=comm.bcast(n, root=MASTER)        
    # Envia el array entero a los workers, lo necesitan entero
    a=comm.bcast(a, root=MASTER)  
    

    # Comienza el timer una vez inicializado todo
    timeStart = MPI.Wtime()
    
    if myrank==MASTER:
        # Init
        arrayProc=0;        
        
        if numWorkers<=n:  
            for i in range(1, numWorkers+1):                           
                # Envia val
                comm.send(a[arrayProc], dest=i)                
                # update
                arrayProc+=1        
        else:
            aux=numWorkers%n
            for i in range(1, numWorkers-aux+1):                      
                # Envia val
                comm.send(a[arrayProc], dest=i)   
                # update
                arrayProc+=1							
            for i in range(numWorkers-aux+1, numWorkers+1):
                comm.send(END_OF_PROCESSING, dest=i)                  
            
            numWorkers-=aux        

        # Hay mas elementos a procesar
        while arrayProc<n:		
            pos=-1;	
            # Recibe los resultados de algun worker
            pos = comm.recv(source=MPI.ANY_SOURCE, tag=tag,status=status)
            source_rank = status.Get_source()
            
            val = comm.recv(source=source_rank, tag=tag,status=status)
            #print("Master recibe:",pos,"de:", source_rank)
            

            # Procesa los datos recibidos
            # Si la posicion en el array b esta ocupado, es porque es el mismo valor
            # y busca el siguiente espacio disponible           
            while b[pos]!=INF: pos+=1
            b[pos]=val

            # Envia el siguiente valor a procesar    
            comm.send(a[arrayProc], dest=status.source)   
            # Update
            arrayProc+=1
            #if arrayProc%(n//10)==0: print("10%",MPI.Wtime()-timeStart)

            # Fin
            if arrayProc==n: break	

            					
        

        # Ultimos valores por procesar
        while numWorkers!=0:
            pos = comm.recv(source=MPI.ANY_SOURCE, tag=tag,status=status)
            source_rank = status.Get_source()
            
            val = comm.recv(source=source_rank, tag=tag,status=status)


            while b[pos]!=INF: pos+=1
            b[pos]=val

            comm.send(END_OF_PROCESSING, dest=status.source, tag=tag)   
            numWorkers-=1
        
        timeEnd = MPI.Wtime()

        print("Tiempo de ejecucion:", timeEnd-timeStart)

        if arrayOrdenado(b,n): print("Array ordenado")
        else: print("Array no ordenado")
    		
    else: # workers
        while(True):
            # Recibe el valor del array a procesar
            val=comm.recv(source=0)

            if val==-2: break

            cont=0
            for i in range(n):            
                if a[i]<val: cont+=1
            
            comm.send(cont, dest=0)
            comm.send(val, dest=0)
            
          
    # End MPI environment
    #MPI.Finalize()

    return 0





def leeArchivo():
    """
    return:
    array: int[].   Array con los enteros leidos
    tam: int.       Tamaño del array leido
    """
    
    dir=os.getcwd()
    n=len(dir)

    while(dir[n-3]!='T' and dir[n-2]!='F' and dir[n-1]!='G'):
        dir=os.path.dirname(dir)
        n=len(dir)

    nombre_fichero=input("Introduce un nombre del fichero: ")    
    path=os.path.join(dir, ".Otros","ficheros","No_Ordenado", nombre_fichero+".txt")
    
       
    tam=0    
    array = [] 
    try:        
        with open(path, 'r') as archivo: # modo lectura
            for linea in archivo: # Solo hay una linea                
                numeros_en_linea = linea.split() # Divide por espacios                               
                for numero in numeros_en_linea:
                    array.append(int(numero))
                    tam+=1
    
    except FileNotFoundError:
        print("El archivo '{}' no existe.".format(nombre_fichero+".txt"))
    
    return array, tam

   


def arrayOrdenado(a, n):
    """
    a: int[]    El array a comprobar
    n: int      Tamaño del array
    return.     True or False
    """
    for i in range(1,n):    
        if (a[i]<a[i-1]):return False
    
    return True


main()